src/main/java/frc/robot/subsystems/Conveyor.java

```diff
@@ -57,94 +113,268 @@ public void setBackMotorPower(double power) {
+
+            if (colorSensorMeasurement != previousColorSensorMeasurement
+                && MIN_COLOR_SENSOR_DISTANCE <= colorSensorDistance
+                && colorSensorDistance <= MAX_COLOR_SENSOR_DISTANCE) {
+                // If we detect a different measurement and it seems valid...
+
+                if ((colorSensorMeasurement == "Blue"
+                    && beltState == ConveyorState.BLUE)
+                    || (colorSensorMeasurement == "Red"
+                    && beltState == ConveyorState.RED)) {
+                    // If that measurement matches beltState...
+
+                    beltState = ConveyorState.NONE;
+
+                } else if (beltState == ConveyorState.NONE
+                    && ((colorSensorMeasurement == "Red"
+                    && indexState == ConveyorState.RED
+                    && !isBallIndexed)
+                    || (colorSensorMeasurement == "Blue"
+                    && indexState == ConveyorState.BLUE
+                    && !isBallIndexed))) {
+                    // If that measurement matches indexState, the ball is not yet indexed, and
+                    // beltState is currently empty...
+
+                    indexState = ConveyorState.NONE;
+
+                }
+            }
+            
+            if (indexRPM > 0
+                && indexState != ConveyorState.NONE
+                && wasBallIndexed
+                && !isBallIndexed) {
+                // If we probably fired the indexed ball...
+
+                indexState = ConveyorState.NONE;
+
+            }
+
+        } else {
+
+            if (indexRPM > 0
+                && indexState != ConveyorState.NONE
+                && wasBallIndexed
+                && !isBallIndexed) {
+                // If the belt is stationary, but we probably still fired the indexed ball...
+
+                indexState = ConveyorState.NONE;
+
+            }
+        }
+    }
+
+    /**
+     * Returns the type of ball present in the position closer to the intake.
+     *
+     * @return the type of ball present in the position closer to the intake
+     */
+    public ConveyorState getBeltState() {
+        return beltState;
+    }
+
+    /**
+     * Returns the type of ball present in the position further away from the intake.
+     * Note: To keep the logic simple, the conveyor code automatically assigns the
+     * furthest-along ball to the indexed state, even if it has not gotten off the conveyor yet.
+     * Check isBallIndexed to see if the ball is actually ready.
+     *
+     * @return the type of ball present in the position further away from the intake
+     */
+    public ConveyorState getWillBeIndexedState() {
+        return indexState;
+    }
+
+    /**
+     * Returns whether or not there is a ball ready to be fired.
+     * This should be checked by the shooter.
+     *
+     * @return whether or not there is a ball ready to be fired
+     */
+    public boolean getIsBallIndexed() {
```

Comment:
pass lidar distance into these so a previous but unenforced distance update is not nessary